src/scipp/coords.py

```diff
@@ -2,247 +2,480 @@
+            return {self.out_names[0]: output}
+        return output
+
+    @property
+    def dependencies(self) -> Tuple[str]:
+        return self._arg_names
+
+    @property
+    def func_name(self) -> str:
+        return self._func.__name__
+
+    def __str__(self):
+        return f'Compute {self._format_out_names()} = {self._func.__name__}' \
+               f'({", ".join(self._arg_names)})'
+
+
+def _make_rule(products, producer) -> _Rule:
+    if isinstance(producer, str):
+        return _RenameRule(products, producer)
+    return _ComputeRule(products, producer)
+
+
+def _non_duplicate_rules(rules: Mapping[str, _Rule]) -> List[_Rule]:
+    already_used = set()
+    result = []
+    for rule in filter(lambda r: r not in already_used,
+                       map(lambda n: rules[n], _sort_topologically(rules))):
+        already_used.add(rule)
+        result.append(rule)
+    return result
+
+
+def _convert_to_rule_graph(graph: GraphDict) -> Dict[str, _Rule]:
+    rule_graph = {}
+    for products, producer in graph.items():
+        products = (products, ) if isinstance(products, str) else tuple(products)
+        rule = _make_rule(products, producer)
+        for product in products:
+            if product in rule_graph:
+                raise ValueError(
+                    f'Duplicate output name defined in conversion graph: {product}')
+            rule_graph[product] = rule
+    return rule_graph
+
+
+def _sort_topologically(graph: Mapping[str, _Rule]) -> Iterable[str]:
+    yield from TopologicalSorter(
+        {out: rule.dependencies
+         for out, rule in graph.items()}).static_order()
+
+
+def _is_meta_data(name: str, data: DataArray) -> bool:
+    return name in data.meta or (data.bins is not None and name in data.bins.meta)
+
+
+def _make_digraph(*args, **kwargs):
+    try:
+        from graphviz import Digraph
+    except ImportError:
+        raise RuntimeError('Failed to import `graphviz`, please install `graphviz` if '
+                           'using `pip`, or `python-graphviz` if using `conda`.')
+    return Digraph(*args, **kwargs)
 
-    def __contains__(self, name):
-        return name in self._graph
 
-    def show(self, size=None, simplified=False):
+class Graph:
+    def __init__(self, graph):
+        self._graph = _convert_to_rule_graph(graph)
+
+    def subgraph(self, data: DataArray, targets: Set[str]) -> Dict[str, _Rule]:
+        subgraph = {}
+        pending = list(targets)
+        while pending:
+            out_name = pending.pop()
+            if out_name in subgraph:
+                continue
+            rule = self._rule_for(out_name, data)
+            subgraph[out_name] = rule
+            pending.extend(rule.dependencies)
+        return subgraph
+
+    def _rule_for(self, out_name: str, data: DataArray) -> _Rule:
+        if _is_meta_data(out_name, data):
+            return _FetchRule((out_name, ), data.meta,
+                              data.bins.meta if data.bins else None)
```

Comment:
May be slightly simpler above we we pass `{}` instead of `None`?

Reply:
God idea!